lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+  case ParseKind::Name: specKind = SpecifierKind::Name; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Position: specKind = SpecifierKind::Position; break;
+  case ParseKind::Read: specKind = SpecifierKind::Read; break;
+  case ParseKind::Readwrite: specKind = SpecifierKind::Readwrite; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sequential: specKind = SpecifierKind::Sequential; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  case ParseKind::Status: specKind = SpecifierKind::Status; break;
+  case ParseKind::Stream: specKind = SpecifierKind::Stream; break;
+  case ParseKind::Unformatted: specKind = SpecifierKind::Unformatted; break;
+  case ParseKind::Write: specKind = SpecifierKind::Write; break;
+  case ParseKind::Convert: specKind = SpecifierKind::Convert; break;
+  case ParseKind::Dispose: specKind = SpecifierKind::Dispose; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::InquireSpec::IntVar &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::InquireSpec::IntVar::Kind;
+  switch (std::get<parser::InquireSpec::IntVar::Kind>(spec.t)) {
+  case ParseKind::Iostat: specKind = SpecifierKind::Iostat; break;
+  case ParseKind::Nextrec: specKind = SpecifierKind::Nextrec; break;
+  case ParseKind::Number: specKind = SpecifierKind::Number; break;
+  case ParseKind::Pos: specKind = SpecifierKind::Pos; break;
+  case ParseKind::Recl: specKind = SpecifierKind::Recl; break;
+  case ParseKind::Size: specKind = SpecifierKind::Size; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::InquireSpec::LogVar &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::InquireSpec::LogVar::Kind;
+  switch (std::get<parser::InquireSpec::LogVar::Kind>(spec.t)) {
+  case ParseKind::Exist: specKind = SpecifierKind::Exist; break;
+  case ParseKind::Named: specKind = SpecifierKind::Named; break;
+  case ParseKind::Opened: specKind = SpecifierKind::Opened; break;
+  case ParseKind::Pending: specKind = SpecifierKind::Pending; break;
+  }
+  SetSpecifier(specKind);
+}
+
+void IoChecker::Enter(const parser::IoControlSpec &spec) {
+  // IoControlSpec context Name
+  hasIoControlList_ = true;
+  if (const parser::Name * name{std::get_if<parser::Name>(&spec.u)}) {
+    CHECK(name->symbol != nullptr && name->symbol->has<NamelistDetails>());
+    SetSpecifier(SpecifierKind::Nml);
+  }
+}
+
+void IoChecker::Enter(const parser::IoControlSpec::Asynchronous &spec) {
+  SetSpecifier(SpecifierKind::Asynchronous);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      hasAsynchronousYes_ = parser::ToUpperCaseLetters(**charConst) == "YES"s;
```

Comment:
This pattern of GetExpr, Fold, Unwrap, ToUpperCaseLetters appears several times. It might be worthwhile extracting it into a function.

Reply:
This is now template function GetConstExpr in check-io.h.  (Thanks for the help writing this!)